core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/common/iterator/AbstractOriginalChangeIterator.java

```diff
@@ -34,7 +37,11 @@ protected Move_ createUpcomingSelection() {
             upcomingEntity = entityIterator.next();
             valueIterator = valueSelector.iterator(upcomingEntity);
         }
+        Object existingValue = entityVariableDescriptor.getValue(upcomingEntity);
         Object toValue = valueIterator.next();
+        if (Objects.equals(existingValue, toValue)) {
+            return createUpcomingSelection();
```

Comment:
This will throw a StackOverflowException in some cases where it shouldn't.

Usless moves return isDoable() = false, and aren't evaluated.